internal/website/gatherexamples/gatherexamples.go

```diff
@@ -0,0 +1,271 @@
+		os.Exit(1)
+	}
+	data = append(data, '\n')
+
+	output := os.Stdout
+	if *outputPath != "" && *outputPath != "-" {
+		var err error
+		output, err = os.Create(*outputPath)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "gatherexamples: %v\n", err)
+			os.Exit(1)
+		}
+		defer func() {
+			if err := output.Close(); err != nil {
+				fmt.Fprintf(os.Stderr, "gatherexamples: write output: %v\n", err)
+				os.Exit(1)
+			}
+		}()
+	}
+	if _, err := output.Write(data); err != nil {
+		fmt.Fprintf(os.Stderr, "gatherexamples: write output: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+const gatherLoadMode packages.LoadMode = packages.NeedName |
+	packages.NeedFiles |
+	packages.NeedTypes |
+	packages.NeedSyntax |
+	packages.NeedTypesInfo
+
+func gather(pkgs []*packages.Package) map[string]string {
+	examples := make(map[string]string)
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Syntax {
+			for _, decl := range file.Decls {
+				// Determine whether this declaration is an example function.
+				fn, ok := decl.(*ast.FuncDecl)
+				if !ok {
+					continue
+				}
+				if !ok || !strings.HasPrefix(fn.Name.Name, "Example") || len(fn.Type.Params.List) > 0 || len(fn.Type.Params.List) > 0 {
+					continue
+				}
+
+				// Gather sorted list of imported packages.
+				usedPackages := make(map[string]string)
+				ast.Inspect(fn.Body, func(node ast.Node) bool {
+					id, ok := node.(*ast.Ident)
+					if !ok {
+						return true
+					}
+					refPkg, ok := pkg.TypesInfo.ObjectOf(id).(*types.PkgName)
+					if !ok {
+						return true
+					}
+					overrideName := ""
+					if id.Name != refPkg.Imported().Name() {
+						overrideName = id.Name
+					}
+					usedPackages[refPkg.Imported().Path()] = overrideName
+					return true
+				})
+
+				// Format example into string.
+				sb := new(strings.Builder)
+				err := printer.Fprint(sb, pkg.Fset, &printer.CommentedNode{
+					Node:     fn.Body,
+					Comments: file.Comments,
+				})
+				if err != nil {
+					panic(err) // will only occur for bad invocations of Fprint
+				}
+				exampleCode := rewriteBlock(sb.String())
+				if len(usedPackages) > 0 {
+					exampleCode = formatImports(usedPackages) + "\n\n" + exampleCode
+				}
+
+				pkgPath := strings.TrimSuffix(pkg.PkgPath, "_test")
+				exampleName := pkgPath + "." + fn.Name.Name
+				examples[exampleName] = exampleCode
+			}
+		}
+	}
+	return examples
+}
+
+// rewriteBlock reformats a Go block statement for display as an example.
+func rewriteBlock(block string) string {
+	// Trim block.
+	block = strings.TrimPrefix(block, "{")
+	block = strings.TrimSuffix(block, "}")
+
+	// Rewrite line-by-line.
+	sb := new(strings.Builder)
+rewrite:
+	for len(block) > 0 {
+		var line string
+		line, block = nextLine(block)
+
+		// Dedent line.
+		// TODO(light): In the case of a multi-line raw string literal,
+		// this can produce incorrect rewrites.
+		line = strings.TrimPrefix(line, "\t")
+
+		// Check if this is a special line.
```

Comment:
Can you explain what "special" means here (or earlier?)

Reply:
Done.